crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -109,37 +109,35 @@ pub unsafe trait SystemParamState: Send + Sync + 'static {
     fn new_archetype(&mut self, _archetype: &Archetype, _system_meta: &mut SystemMeta) {}
     #[inline]
     fn apply(&mut self, _world: &mut World) {}
-}
-
-/// A [`SystemParamFetch`] that only reads a given [`World`].
-///
-/// # Safety
-/// This must only be implemented for [`SystemParamFetch`] impls that exclusively read the World passed in to [`SystemParamFetch::get_param`]
-pub unsafe trait ReadOnlySystemParamFetch {}
 
-pub trait SystemParamFetch<'world, 'state>: SystemParamState {
-    type Item: SystemParam<Fetch = Self>;
+    type Item<'world, 'state>: SystemParam<State = Self>;
     /// # Safety
     ///
     /// This call might access any of the input parameters in an unsafe way. Make sure the data
     /// access is safe in the context of the system scheduler.
-    unsafe fn get_param(
+    unsafe fn get_param<'world, 'state>(
         state: &'state mut Self,
         system_meta: &SystemMeta,
         world: &'world World,
         change_tick: u32,
-    ) -> Self::Item;
+    ) -> Self::Item<'world, 'state>;
 }
 
+/// A [`SystemParam`] that only reads a given [`World`].
+///
+/// # Safety
+/// This must only be implemented for [`SystemParam`] impls that exclusively read the World passed in to [`SystemParamState::get_param`]
+pub unsafe trait ReadOnlySystemParam {}
```

Comment:
This should be a subtrait of `SystemParam`. The same way `ReadOnlyWorldQuery: WorldQuery`.

Reply:
I left that out, because it causes inscrutable type errors like this:

```
error[E0310]: the parameter type `Q` may not live long enough
   --> crates\bevy_ecs\src\system\system_param.rs:139:67
    |
139 | unsafe impl<'w, 's, Q: ReadOnlyWorldQuery, F: ReadOnlyWorldQuery> ReadOnlySystemParam
    |                                                                   ^^^^^^^^^^^^^^^^^^^ ...so that the type `Q` will meet its required lifetime bounds...
    |
note: ...that is required by this bound
   --> crates\bevy_ecs\src\system\system_param.rs:130:39
    |
130 | pub unsafe trait ReadOnlySystemParam: SystemParam {}
    |                                       ^^^^^^^^^^^
help: consider adding an explicit lifetime bound...
    |
139 | unsafe impl<'w, 's, Q: ReadOnlyWorldQuery + 'static, F: ReadOnlyWorldQuery> ReadOnlySystemParam
    |                                           +++++++++
```

For some reason, adding `SystemParam` as a supertrait causes rustc to think these parameters must be `'static`? Maybe I'm just missing something, but this doesn't seem to work.

Reply:
That looks correct to me given the constraints on the other `StaticSystemParam*` traits/types.

Reply:
Oh I see, the error message is just pointing to the wrong bound.